README.md
# OpenExchangeRates
[![Build Status](https://semaphoreci.com/api/v1/smeevil/open_exchange_rates/branches/master/shields_badge.svg)](https://semaphoreci.com/smeevil/open_exchange_rates) [![Coverage Status](https://coveralls.io/repos/github/smeevil/open_exchange_rates/badge.svg?branch=master)](https://coveralls.io/github/smeevil/open_exchange_rates?branch=master) [![Deps Status](https://beta.hexfaktor.org/badge/all/github/smeevil/open_exchange_rates.svg)](https://beta.hexfaktor.org/github/smeevil/open_exchange_rates)

This Elixir library uses the [openexchangerates.org](https://openexchangerates.org) API to retieve the latest exchange rates.

At initialization of the library, it will use the latest exchange rates from a cache file. After initialization it will immediately try to get the latest exchange rates from openexchangerates.org. If this fails it will retry to get a new update every minute.

After a successful update, it will check every minute if the cache needs to be updated and fetches the new rates from openexchangerates.org. You can configure the cache time, which by default is 24 hours. Please take in account that every check will be taking credits from your API usage.

To be able to use this library you will need an API token from openexchangerates.org which you can get [here](https://openexchangerates.org/signup).

## Using it without an API key
This library will still function without a connection to the [openexchangerates.org](https://openexchangerates.org) API but please take into account the exchanges rates are coming from an (outdated) cache. This cache will be updated with every release of this library.
## Example usage
This library gives you the following functions :

- List available currencies
```elixir
OpenExchangeRates.available_currencies |> Enum.take(10)
["AWG", "NAD", "INR", "LAK", "BOB", "MOP", "QAR", "SDG", "TMT", "BRL"]
```

- Get the exchange rate for USD to an other currency
```elixir
OpenExchangeRates.rate_for_currency(:EUR)
{:ok, 0.902}
```

- Convert any currency to an other
```elixir
OpenExchangeRates.convert(100.00, :EUR, :GBP)
{:ok, 84.81186252771619}

OpenExchangeRates.convert!(100.00, :EUR, :GBP)
84.81186252771619
```

- Convert cents in any currency to an other
```elixir
OpenExchangeRates.convert_cents(100, :GBP, :AUD)
{:ok, 172}

OpenExchangeRates.convert_cents!(100, :GBP, :AUD)
172
```

- convert a currency and return a properly formatted string for that currency
```elixir
OpenExchangeRates.convert_and_format(1234, :EUR, :AUD)
{:ok, "A$1,795.10"}

OpenExchangeRates.convert_and_format!(1234, :EUR, :AUD)
"A$1,795.10"
```

- convert cents and return a properly formatted string for that currency
```elixir
OpenExchangeRates.convert_cents_and_format(1234567, :EUR, :NOK)
{:ok, "116.495,78NOK"}

OpenExchangeRates.convert_cents_and_format!(1234567, :EUR, :NOK)
"116.495,78NOK"
```

- Get the age of the cache in seconds
```elixir
OpenExchangeRates.cache_age
25341
```
## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed as:

  1. Add `open_exchange_rates` to your list of dependencies in `mix.exs`:

    ```elixir
    def deps do
      [{:open_exchange_rates, "~> 0.1.0"}]
    end
    ```

  2. Ensure `open_exchange_rates` is started before your application:

    ```elixir
    def application do
      [applications: [:open_exchange_rates]]
    end
    ```

## Configuration

Please add the following config to your config.exs
```elixir
config :open_exchange_rates,
  app_id: "MY API KEY",
  cache_time_in_minutes: 1440 #24 hours
```


## Testing the library
Before you run the tests, please make sure to set the OER_APP_ID environment to you app_id key.



CHANGELOG.md
## 0.3.0 (2016-08-12)
  - The updater is no longer a genserver but a recursive call which is started and linked by the cache
  - Cache can now return a cache age and the updater will properly honer th cache time.
  - When starting the app, it will check the cache age of the cache file, and use that if its still valid, otherwise do a direct update
  - Added bang methods
  - Added methods to return a formatted currency string

## 0.2.0 (2016-08-08)
  - OpenExchangeRates.rate_for_currency/1 is now OpenExchangeRates.conversion_rate/2
  - Config now accepts two new settings
    - cache_file: "/path/to/cache.json"
    - auto_update: true
  - Supervisor now is a one_for_all, in stead of one_for_one, this makes sure that the updater and cache are linked to each other.
  - bumped Credo version

## 0.1.0 (2016-08-07)

  - Initial release


config/test.exs
use Mix.Config

config :exvcr, [
  vcr_cassette_library_dir: "fixture/vcr_cassettes",
  custom_cassette_library_dir: "fixture/custom_cassettes",
  filter_sensitive_data: [
    [pattern: "app_id=.*", placeholder: "app_id=APP_ID_PLACEHOLDER"]
  ],
  filter_url_params: false,
  response_headers_blacklist: []
]

config :open_exchange_rates,
  app_id: System.get_env("OER_APP_ID"),
  cache_time_in_minutes: 1440,
  cache_file: File.cwd! <> "/fixture/exchange_data.json",
  auto_update: false


test/open_exchange_rates_test.exs
defmodule OpenExchangeRatesTest do
  use ExUnit.Case
  doctest OpenExchangeRates

  test "currencies" do
    assert 172 = OpenExchangeRates.available_currencies |> Enum.count
  end

  test "convert non existing currency" do
    assert {:error, "unknown currency: BLA"} = OpenExchangeRates.convert_cents(100, :BLA, :EUR)
  end

  test "get conversion rate between two currencies" do
    {:ok, 0.8481186252771619} = OpenExchangeRates.conversion_rate(:EUR, :GBP)
  end

  test "convert usd to eur" do
    assert  {:ok, 90} = OpenExchangeRates.convert_cents(100, :USD, :EUR)
    assert  {:ok,  0.902} = OpenExchangeRates.convert(1.0, :USD, :EUR)
  end

  test "convert eur to usd" do
    assert  {:ok, 111} = OpenExchangeRates.convert_cents(100, :EUR, :USD)
    assert  {:ok, 1.1086474501108647} = OpenExchangeRates.convert(1.0, :EUR, :USD)
  end

  test "convert eur to gbp" do
    assert  {:ok, 85} = OpenExchangeRates.convert_cents(100, :EUR, :GBP)
    assert   {:ok, 0.8481186252771619} = OpenExchangeRates.convert(1.0, :EUR, :GBP)
  end

  test "convert gbp to eur" do
    assert  {:ok, 118} = OpenExchangeRates.convert_cents(100, :GBP, :EUR)
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.convert(1.00, :GBP, :EUR)
  end


  test "sanitize user input" do
    assert  {:ok, 118} = OpenExchangeRates.convert_cents(100, "GBP", "EUR")
    assert  {:ok, 118} = OpenExchangeRates.convert_cents(100, "gbp", "eur")
    assert  {:ok, 118} = OpenExchangeRates.convert_cents(100, :gbp, :eur)
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.convert(1, "GBP", "EUR")
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.convert(1.0, "GBP", "EUR")
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.convert(1.0, "gbp", "eur")
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.convert(1.0, :gbp, :eur)
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.conversion_rate(:gbp, :eur)
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.conversion_rate("gbp", "eur")
    assert  {:ok, 1.1790803434757773} = OpenExchangeRates.conversion_rate("GBP", "EUR")
  end

  test "it should return the cache age" do
    assert is_integer(OpenExchangeRates.cache_age)
  end

  test "it should return an error when formatting an unknown currency" do
    assert {:error, "unknown currency: BLA"} = OpenExchangeRates.convert_and_format(123456789, :EUR, :BLA)
  end
  test "it should convert and return a formatted string" do
    assert {:ok, "£104,706,002.17"} = OpenExchangeRates.convert_and_format(123456789, :EUR, :GBP)
    assert "£104,706,002.17" = OpenExchangeRates.convert_and_format!(123456789, :EUR, :GBP)

    assert {:ok, "£1,047,060.02"} = OpenExchangeRates.convert_cents_and_format(123456789, :EUR, :GBP)
    assert "£1,047,060.02" = OpenExchangeRates.convert_cents_and_format!(123456789, :EUR, :GBP)
  end

  test "it show raise when convert! with unknown currency" do
    assert_raise RuntimeError, "unknown currency: BLA", fn -> OpenExchangeRates.convert!(100, :EUR, :BLA) end
  end

  test "it show raise when convert_cents! with unknown currency" do
    assert_raise RuntimeError, "unknown currency: BLA", fn -> OpenExchangeRates.convert_cents!(100, :EUR, :BLA) end
  end

  test "it show raise when convert_and_format! with unknown currency" do
    assert_raise RuntimeError, "unknown currency: BLA", fn -> OpenExchangeRates.convert_and_format!(100, :EUR, :BLA) end
  end

  test "it show raise when convert_cents_and_format! with unknown currency" do
    assert_raise RuntimeError, "unknown currency: BLA", fn -> OpenExchangeRates.convert_cents_and_format!(100, :EUR, :BLA) end
  end
end


test/open_exchange_rates/cache_test.exs
defmodule OpenExchangeRates.CacheTest do
  use ExUnit.Case
  use ExVCR.Mock, adapter: ExVCR.Adapter.Hackney

  doctest OpenExchangeRates.Cache

  test "it should an error for an unknown currency" do
    assert  {:error, "unknown currency: BLA"} == OpenExchangeRates.Cache.rate_for_currency("BLA")
  end

  test "it should return a rate for a currency" do
    assert  {:ok, 0.902} == OpenExchangeRates.Cache.rate_for_currency("EUR")
  end

  test "it should update its rates" do
    {:ok, rate} = OpenExchangeRates.Cache.rate_for_currency("EUR")
    assert 0.902 == rate
    OpenExchangeRates.Cache.update!(%{"EUR" => 0.1}, :os.system_time(:seconds))
    assert {:ok, 0.1} == OpenExchangeRates.Cache.rate_for_currency("EUR")

    #reset to original value
    OpenExchangeRates.Cache.update!(%{"EUR" => rate}, :os.system_time(:seconds))
  end
end


mix.exs
defmodule OpenExchangeRates.Mixfile do
  use Mix.Project

  def project do
    [
      app: :open_exchange_rates,
      description: "A library that can convert currencies and uses automatic updating of the currrency rates using openexchangerates.org",
      package: package,
      version: "0.3.0",
      elixir: "~> 1.3",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      docs: [extras: ["README.md", "CHANGELOG.md"]],
      preferred_cli_env: ["coveralls": :test, "coveralls.detail": :test, "coveralls.post": :test, "coveralls.html": :test],
      test_coverage: [tool: ExCoveralls]
    ]
  end

  def application do
    [
      mod: {OpenExchangeRates, []},
      applications: [:logger, :httpoison, :currency_formatter]
    ]
  end

  defp deps do
    [
      {:httpoison, "~>0.9.0"},
      {:poison, "~>2.2.0"},
      {:exvcr, "~>0.8.1", only: [:dev, :test]},
      {:credo, "~>0.4.7", only: [:dev, :test]},
      {:mix_test_watch, "~>0.2.6", only: [:dev, :test]},
      {:ex_doc, "~>0.13.0", only: [:dev, :test]},
      {:excoveralls, "~> 0.5", only: :test},
      {:currency_formatter, "~>0.4.2"}
    ]
  end

  defp package do
    [
      maintainers: ["Gerard de Brieder"],
      licenses: ["MIT"],
      files: ["lib", "priv", "mix.exs", "README*", "LICENSE*", "CHANGELOG*"],
      links: %{
        "GitHub" => "https://github.com/smeevil/open_exchange_rates",
        "Docs"   => "https://smeevil.github.com/open_exchange_rates",
        "Changelog"   => "https://smeevil.github.com/open_exchange_rates/changelog.html",
      }
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :open_exchange_rates, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:open_exchange_rates, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#

import_config "#{Mix.env}.exs"


lib/open_exchange_rates/cache.ex
defmodule OpenExchangeRates.Cache do
  @moduledoc """
  This module takes care of caching the currency rates and is for internal use only
  """
  use GenServer


  @doc false
  def start_link(configuration_status) do
    GenServer.start_link(__MODULE__, configuration_status, name: __MODULE__)
  end

  @doc false
  def init(configuration_status) do
    ets = :ets.new(__MODULE__, [:set, :protected, :named_table])
    start_updater(configuration_status)
    {:ok, %{ets: ets, cached_at: 0}}
  end

  @doc """
  Get the age of the cache in seconds
  """
  @spec cache_age() :: Integer.t
  def cache_age do
    cached_at = GenServer.call(__MODULE__, {:cached_at})
    :os.system_time(:seconds) - cached_at
  end

  @doc false
  @spec rate_for_currency((String.t| Atom.t)) :: {:ok, Float.t} | {:error, String.t}
  def rate_for_currency(currency) when is_atom(currency), do: rate_for_currency(Atom.to_string(currency))
  def rate_for_currency(currency) when is_binary(currency) do
    case OpenExchangeRates.Cache |> :ets.lookup(String.upcase(currency)) do
      [] -> {:error, "unknown currency: #{currency}"}
      [{_, rate}] -> {:ok, rate}
    end
  end

  @doc false
  @spec currencies() :: [String.t]
  def currencies, do: OpenExchangeRates.Cache |> :ets.match({:"$1", :"_"}) |> List.flatten

  @doc false
  @spec update!([{String.t, Float.t}], Integer.t) :: :ok
  def update!(rates, cached_at) do
    GenServer.call(__MODULE__, {:update!, rates, cached_at})
  end

  @doc false
  def handle_call({:update!, rates, cached_at}, _caller, %{ets: ets} = state) do
    Enum.each(rates, fn {currency, rate} -> :ets.insert(ets, {currency, rate}) end)
    state = Map.put(state, :cached_at, cached_at)
    {:reply, :ok, state}
  end

  @doc false
  def handle_call({:cached_at}, _caller, %{cached_at: cached_at} = state), do: {:reply, cached_at, state}


  def start_updater(configuration_status) do
    function = case configuration_status do
      :disable_updater -> &OpenExchangeRates.Updater.load_data_from_disk!/0
      :missing_key -> &OpenExchangeRates.Updater.load_data_from_disk!/0
      _other -> &OpenExchangeRates.Updater.start/0
    end
    spawn_link(function)
  end
end


lib/open_exchange_rates/client.ex
defmodule OpenExchangeRates.Client do
  @moduledoc """
  This module takes care of the API communication between openexchangerates.org and is for internal use only
  """
  require Logger

  @app_id Application.get_env(:open_exchange_rates, :app_id)

  @doc false
  @spec get_latest() :: {:ok, Map.t} | {:error, String.t}
  def get_latest do
    response = HTTPoison.get("https://openexchangerates.org/api/latest.json?base=USD&app_id=#{@app_id}")
    case response do
      {:ok, %HTTPoison.Response{body: json, status_code: 200}} -> json |> handle_api_success
      {:ok, %HTTPoison.Response{body: json}} -> json |> handle_api_error
      {:error, _} = error -> error
    end
  end

  defp handle_api_success(json) do
    data = Poison.decode!(json)
    {:ok, data}
  end

  defp handle_api_error(body) do
    case Poison.decode(body) do
      {:ok, data} ->
        if data["message"] == "invalid_app_id", do: Logger.error("OpenExchangeRates: The app id `#{@app_id}` is not valid !")
        {:error, data["description"]}
      _ -> {:error, "Could not parse the JSON : #{inspect body}"}
    end
  end
end


lib/open_exchange_rates/updater.ex
defmodule OpenExchangeRates.Updater do
  @moduledoc """
  This module takes care of updating the cache with new conversion rates from openexchangerates.org and is for internal use only.
  """

  require Logger

  @update_interval_in_seconds 60

  def cache_file, do: Application.get_env(:open_exchange_rates, :cache_file) || (File.cwd! <> "/priv/latest.json")

  @doc false
  def start do
    load_data_from_disk
    check_for_update
    :ok
  end

  @doc false
  def check_for_update do
    cache_time = Application.get_env(:open_exchange_rates, :cache_time_in_minutes, 1440) * 60

    if OpenExchangeRates.Cache.cache_age >= cache_time, do: update
    Process.sleep(@update_interval_in_seconds * 1000)
    check_for_update
  end

  def load_data_from_disk do
    if File.exists?(cache_file()) do
      cached_at = File.lstat!(cache_file(), [time: :posix]).mtime
      cache_file()
      |> File.read!
      |> Poison.decode!
      |> update_cache!(cached_at)
      :ok
    else
      {:error, "could not find #{cache_file()} which you defined in your config"}
    end
  end

  def load_data_from_disk! do
    case load_data_from_disk do
      :ok -> :ok
      {:error, message} -> raise(message)
    end
  end

  defp write_to_disk!(data) do
    json = Poison.encode!(data)
    File.write!(cache_file(), json)
    data
  end

  defp update do
    Logger.info "OpenExchangeRates Updating Rates..."
    case OpenExchangeRates.Client.get_latest do
      {:ok, data} -> data |> write_to_disk! |> update_cache!
      {:error, message} -> {:error, message}
    end
  end

  defp update_cache!(data, cached_at \\ nil)
  defp update_cache!(%{"rates" => rates}, cached_at) do
    cached_at = cached_at || :os.system_time(:seconds)
    GenServer.call(OpenExchangeRates.Cache, {:update!, rates, cached_at})
  end
  defp update_cache!(data, _cached_at), do: raise "Data was corrupted ? #{inspect data}"

end


lib/open_exchange_rates.ex
defmodule OpenExchangeRates do
  @moduledoc """
  This module contains all the helper methods for converting currencies
  """
  use Application
  require Logger

  @doc false
  def start(_type, _args) do
    import Supervisor.Spec, warn: false

    configuration_status = check_configuration
    children = [worker(OpenExchangeRates.Cache, [configuration_status])]

    opts = [strategy: :one_for_one, name: OpenExchangeRates.Supervisor]
    Supervisor.start_link(children, opts)
  end

  @doc"""
  Returns a list of all available currencies.

  ## example

      iex> OpenExchangeRates.available_currencies |> Enum.take(10)
      ["AWG", "NAD", "INR", "LAK", "BOB", "MOP", "QAR", "SDG", "TMT", "BRL"]

  """
  @spec available_currencies() :: [String.t]
  def available_currencies, do: OpenExchangeRates.Cache.currencies

  @doc"""
  Returns the age of the cache in seconds

  ## example
      OpenExchangeRates.cache_age
      25341
  """
  @spec cache_age() :: Integer.t
  def cache_age, do: OpenExchangeRates.Cache.cache_age

  @doc"""
  Will convert a price from once currency to another

  ## example

      iex> OpenExchangeRates.convert(100.00, :EUR, :GBP)
      {:ok, 84.81186252771619}

  """
  @spec convert(Integer.t, (String.t | Atom.t), (String.t | Atom.t)) :: {:ok, Float.t} | {:error, String.t}
  def convert(value, from, to) when is_integer(value), do: convert((value/1.0), from, to)
  @spec convert(Float.t, (String.t | Atom.t), (String.t | Atom.t)) :: {:ok, Float.t} | {:error, String.t}
  def convert(value, from, to) when is_float(value) do
    with \
      {:ok, rate_from} <- OpenExchangeRates.Cache.rate_for_currency(from),
      {:ok, rate_to} <- OpenExchangeRates.Cache.rate_for_currency(to) \
    do
      rate_usd = value / rate_from
      converted = rate_usd * rate_to
      {:ok, converted}
    else
      error -> error
    end
  end


  @doc """
  bang method of convert/3
  Will either return the result or raise when there was an error
  ## examples

      iex> OpenExchangeRates.convert!(100.00, :EUR, :GBP)
      84.81186252771619

      iex> OpenExchangeRates.convert!(100, :EUR, :GBP)
      84.81186252771619
  """
  @spec convert!(Integer.t, (String.t | Atom.t), (String.t | Atom.t)) :: {:ok, Float.t} | {:error, String.t}
  def convert!(value, from, to) when is_integer(value), do: convert!((value/1.0), from, to)
  @spec convert!(Float.t, (String.t | Atom.t), (String.t | Atom.t)) :: {:ok, Float.t} | {:error, String.t}
  def convert!(value, from, to) when is_float(value) do
    case convert(value, from, to) do
      {:ok, result} -> result
      {:error, message} -> raise(message)
    end
  end

  @doc"""
  Will convert cents from once currency to another

  ## example

      iex> OpenExchangeRates.convert_cents(100, :GBP, :AUD)
      {:ok, 172}

  """
  @spec convert_cents(Integer.t, (String.t | Atom.t), (String.t | Atom.t)) :: {:ok, Integer.t} | {:error, String.t}
  def convert_cents(value, from, to) when is_integer(value) do
    case convert(value/100, from, to) do
      {:ok, result} -> {:ok, Kernel.round(result * 100)}
      error -> error
    end
  end

  @doc """
  bang method of convert_cents/3
  Will either return the result or raise when there was an error
  ## example

      iex> OpenExchangeRates.convert_cents!(100, :GBP, :AUD)
      172
  """
  @spec convert_cents!(Integer.t, (String.t | Atom.t), (String.t | Atom.t)) :: {:ok, Integer.t} | {:error, String.t}
  def convert_cents!(value, from, to) when is_integer(value) do
    case convert_cents(value, from, to) do
      {:ok, result} -> result
      {:error, message} -> raise(message)
    end
  end

  @doc"""
  Converts cents and returns a properly formatted string for the given currency.

  # Examples

      iex> OpenExchangeRates.convert_cents_and_format(1234567, :EUR, :CAD)
      {:ok, "C$18,026.07"}

      iex> OpenExchangeRates.convert_cents_and_format(1234567, :EUR, :USD)
      {:ok, "$13,687"}

      iex> OpenExchangeRates.convert_cents_and_format(1234567, :USD, :EUR)
      {:ok, "€11.135,79"}

      iex> OpenExchangeRates.convert_cents_and_format(1234567, :EUR, :NOK)
      {:ok, "116.495,78NOK"}
  """
  @spec convert_cents_and_format(Integer.t, (Atom.t | String.t), (Atom.t | String.t)) :: String.t
  def convert_cents_and_format(value, from, to) when is_integer(value) do
    case convert_cents(value, from, to) do
      {:ok, result} -> {:ok, CurrencyFormatter.format(result, to)}
      error -> error
    end
  end

  @doc """
  Bang version of convert_cents_and_format/3
  Will either return the result or raise when there was an error

  #example

      iex> OpenExchangeRates.convert_cents_and_format!(1234567, :EUR, :USD)
      "$13,687"
  """
  @spec convert_cents_and_format!(Integer.t, (Atom.t | String.t), (Atom.t | String.t)) :: String.t
  def convert_cents_and_format!(value, from, to) when is_integer(value) do
    case convert_cents_and_format(value, from, to) do
      {:ok, result} -> result
      {:error, message} -> raise(message)
    end
  end

  @doc"""
  Converts a price and returns a properly formatted string for the given currency.

  # Examples

      iex> OpenExchangeRates.convert_and_format(1234, :EUR, :AUD)
      {:ok, "A$1,795.10"}
  """
  @spec convert_and_format((Integer.t | Float.t), (Atom.t | String.t), (Atom.t | String.t)) :: String.t
  def convert_and_format(value, from, to), do: convert_cents_and_format((Kernel.round(value * 100)), from, to)


  @doc """
  Bang version of convert_and_format/3
  Will either return the result or raise when there was an error

  #example

      iex> OpenExchangeRates.convert_and_format!(1234567, :EUR, :USD)
      "$1,368,699.56"
  """
  @spec convert_and_format!((Integer.t | Float.t), (Atom.t | String.t), (Atom.t | String.t)) :: String.t
  def convert_and_format!(value, from, to) when is_integer(value) do
    case convert_and_format(value, from, to) do
      {:ok, result} -> result
      {:error, message} -> raise(message)
    end
  end

  @doc """
  Get the conversion rate for a between two currencies"

  ## Example

      iex> OpenExchangeRates.conversion_rate(:EUR, :GBP)
      {:ok, 0.8481186252771619}

  """
  @spec conversion_rate((String.t| Atom.t), (String.t| Atom.t)) :: {:ok, Float.t} | {:error, String.t}
  def conversion_rate(from, to) when is_binary(from) and is_binary(to), do: conversion_rate(String.to_atom(from), String.to_atom(to))
  def conversion_rate(from, to), do: convert(1.0, from, to)

  defp check_configuration do
    cond do
      Application.get_env(:open_exchange_rates, :auto_update) == false -> :disable_updater
      Application.get_env(:open_exchange_rates, :app_id) == nil -> config_error_message; :missing_key
      true -> :ok
    end
  end

  defp config_error_message do
    Logger.warn ~s[
OpenExchangeRates :

No App ID provided.

Please check if your config.exs contains the following :
  config :open_exchange_rates,
    app_id: "MY_OPENEXCHANGE_RATES_ORG_API_KEY",
    cache_time_in_minutes: 1440,
    cache_file: File.cwd! <> "/priv/exchange_rate_cache.json",
    auto_update: true

If you need an api key please sign up at https://openexchangerates.org/signup

This module will continue to function but will use (outdated) cached exchange rates data...
    ]
  end
end
